<h2><center> German Credit Prediction </center><h>

In [1]:
#importing all the required packages
import pandas as pd
import sklearn as sk
import numpy as np
import os.path

## 1. Loading the dataset from CSV

In [2]:
cwd = os.getcwd()
url = cwd + '/DATA/index.csv'
GermanCredit = pd.read_csv(url)
GermanCredit.head() #TO view the Sample Data

Creditability  Account Balance  Duration of Credit (month)  \
0              1                1                          18   
1              1                1                           9   
2              1                2                          12   
3              1                1                          12   
4              1                1                          12   

   Payment Status of Previous Credit  Purpose  Credit Amount  \
0                                  4        2           1049   
1                                  4        0           2799   
2                                  2        9            841   
3                                  4        0           2122   
4                                  4        0           2171   

   Value Savings/Stocks  Length of current employment  Instalment per cent  \
0                     1                             2                    4   
1                     1                             3                    2   
2                     2                             4                    2   
3                     1                             3                    3   
4                     1                             3                    4   

   Sex & Marital Status       ...        Duration in Current address  \
0                     2       ...                                  4   
1                     3       ...                                  2   
2                     2       ...                                  4   
3                     3       ...                                  2   
4                     3       ...                                  4   

   Most valuable available asset  Age (years)  Concurrent Credits  \
0                              2           21                   3   
1                              1           36                   3   
2                              1           23                   3   
3                              1           39                   3   
4                              2           38                   1   

   Type of apartment  No of Credits at this Bank  Occupation  \
0                  1                           1           3   
1                  1                           2           3   
2                  1                           1           2   
3                  1                           2           2   
4                  2                           2           2   

   No of dependents  Telephone  Foreign Worker  
0                 1          1               1  
1                 2          1               1  
2                 1          1               1  
3                 2          1               2  
4                 1          1               2  

[5 rows x 21 columns]

In [3]:
GermanCredit.columns

Index([u'Creditability', u'Account Balance', u'Duration of Credit (month)',
       u'Payment Status of Previous Credit', u'Purpose', u'Credit Amount',
       u'Value Savings/Stocks', u'Length of current employment',
       u'Instalment per cent', u'Sex & Marital Status', u'Guarantors',
       u'Duration in Current address', u'Most valuable available asset',
       u'Age (years)', u'Concurrent Credits', u'Type of apartment',
       u'No of Credits at this Bank', u'Occupation', u'No of dependents',
       u'Telephone', u'Foreign Worker'],
      dtype='object')

## 2. Data Cleaning Process

In [4]:
GermanCredit.rename(columns = {'Duration of Credit (month)':'CreditDuration'
                              ,'Length of current employment':'CurrentEmploymentDuration'
                              ,'Duration in Current address':'CurrentAddressDuration'
                              ,'No of Credits at this Bank':'NoOfCreditsInBank'
                              ,'Type of apartment':'ApartmentType'
                              ,'Instalment per cent':'InstalmentPercentage'
                              ,'No of dependents': 'NoOfDependents'
                              ,'Most valuable available asset':'MostValuableAsset'
                              ,'Value Savings/Stocks': 'ValueSavingsOrStocks'
                              ,'Sex & Marital Status': 'SexAndMaritalStatus'},inplace=True)

In [5]:
GermanCredit.isnull() #To check if we have any null values in the data

Creditability  Account Balance  CreditDuration  \
0            False            False           False   
1            False            False           False   
2            False            False           False   
3            False            False           False   
4            False            False           False   
5            False            False           False   
6            False            False           False   
7            False            False           False   
8            False            False           False   
9            False            False           False   
10           False            False           False   
11           False            False           False   
12           False            False           False   
13           False            False           False   
14           False            False           False   
15           False            False           False   
16           False            False           False   
17           False            False           False   
18           False            False           False   
19           False            False           False   
20           False            False           False   
21           False            False           False   
22           False            False           False   
23           False            False           False   
24           False            False           False   
25           False            False           False   
26           False            False           False   
27           False            False           False   
28           False            False           False   
29           False            False           False   
..             ...              ...             ...   
970          False            False           False   
971          False            False           False   
972          False            False           False   
973          False            False           False   
974          False            False           False   
975          False            False           False   
976          False            False           False   
977          False            False           False   
978          False            False           False   
979          False            False           False   
980          False            False           False   
981          False            False           False   
982          False            False           False   
983          False            False           False   
984          False            False           False   
985          False            False           False   
986          False            False           False   
987          False            False           False   
988          False            False           False   
989          False            False           False   
990          False            False           False   
991          False            False           False   
992          False            False           False   
993          False            False           False   
994          False            False           False   
995          False            False           False   
996          False            False           False   
997          False            False           False   
998          False            False           False   
999          False            False           False   

     Payment Status of Previous Credit  Purpose  Credit Amount  \
0                                False    False          False   
1                                False    False          False   
2                                False    False          False   
3                                False    False          False   
4                                False    False          False   
5                                False    False          False   
6                                False    False          False   
7                                False    False          False   


In [6]:
GermanCredit.isnull().any()

Creditability                        False
Account Balance                      False
CreditDuration                       False
Payment Status of Previous Credit    False
Purpose                              False
Credit Amount                        False
ValueSavingsOrStocks                 False
CurrentEmploymentDuration            False
InstalmentPercentage                 False
SexAndMaritalStatus                  False
Guarantors                           False
CurrentAddressDuration               False
MostValuableAsset                    False
Age (years)                          False
Concurrent Credits                   False
ApartmentType                        False
NoOfCreditsInBank                    False
Occupation                           False
NoOfDependents                       False
Telephone                            False
Foreign Worker                       False
dtype: bool

The Data does not contain any NaN values, it is clear that all the cells are filled with some values.

In [7]:
GermanCredit.shape

(1000, 21)

In [8]:
GC_with_Condition = GermanCredit.dropna(thresh=1)

In [9]:
GC_with_Condition.shape

(1000, 21)

In [10]:
GC_with_Condition.columns = GC_with_Condition.columns.str.replace(' ','') #Just renaming the column Names

Now, We are good with the data contents. We need to check the irregularity in value distribution in each column for better data analysis

To make sure that all the values in categorical variable are defined clearly, I am exploring the categorical variables using value count function.

In [11]:
GC_with_Condition['SexAndMaritalStatus'].value_counts()

3    548
2    310
4     92
1     50
Name: SexAndMaritalStatus, dtype: int64

In [12]:
GC_with_Condition['ApartmentType'].value_counts()

2    714
1    179
3    107
Name: ApartmentType, dtype: int64

In [13]:
GC_with_Condition['ForeignWorker'].value_counts()

1    963
2     37
Name: ForeignWorker, dtype: int64

In [14]:
GC_with_Condition['MostValuableAsset'].value_counts()

3    332
1    282
2    232
4    154
Name: MostValuableAsset, dtype: int64

In [15]:
GC_with_Condition['Occupation'].value_counts()

3    630
2    200
4    148
1     22
Name: Occupation, dtype: int64

In [16]:
GC_with_Condition['ValueSavingsOrStocks'].value_counts()

1    603
5    183
2    103
3     63
4     48
Name: ValueSavingsOrStocks, dtype: int64

So we are sure that the data is clean and ready for Predictive Analytics,

## 3 : Modeling Process

In [17]:
#Data split with dependent and independent variables in the dataset

X = GC_with_Condition.iloc[:,1:21] #dataset with only dependent variable
Y = GC_with_Condition.iloc[:,0] #dataset with only independent variable

Y.head()

0    1
1    1
2    1
3    1
4    1
Name: Creditability, dtype: int64

In [30]:
from sklearn.cross_validation import train_test_split
GC_X_Train, GC_X_Test, GC_Y_Train, GC_Y_Test = train_test_split(X,Y,test_size=0.2, random_state=0)

In [19]:
GC_X_Train.sample(10)

AccountBalance  CreditDuration  PaymentStatusofPreviousCredit  Purpose  \
353               1              12                              4        0   
132               3               6                              2        2   
639               1              24                              4        1   
591               1               8                              4       10   
548               2              12                              2        0   
487               2              18                              4        2   
309               4              54                              0        1   
849               2              36                              2        6   
798               1              24                              2        0   
179               4              48                              3        3   

     CreditAmount  ValueSavingsOrStocks  CurrentEmploymentDuration  \
353          3499                     1                          3   
132          2116                     1                          3   
639          2957                     1                          5   
591          1164                     1                          5   
548          1007                     4                          3   
487          3612                     1                          5   
309          9436                     5                          3   
849         12612                     2                          3   
798           915                     5                          5   
179         12749                     3                          4   

     InstalmentPercentage  SexAndMaritalStatus  Guarantors  \
353                     3                    2           2   
132                     2                    3           1   
639                     4                    3           1   
591                     3                    3           1   
548                     4                    4           1   
487                     3                    2           1   
309                     2                    3           1   
849                     1                    3           1   
798                     4                    2           1   
179                     4                    3           1   

     CurrentAddressDuration  MostValuableAsset  Age(years)  ConcurrentCredits  \
353                       2                  1          29                  3   
132                       2                  1          41                  3   
639                       4                  2          63                  3   
591                       4                  4          51                  1   
548                       1                  1          22                  3   
487                       4                  2          37                  3   
309                       2                  2          39                  3   
849                       4                  4          47                  3   
798                       2                  3          29                  1   
179                       1                  3          37                  3   

     ApartmentType  NoOfCreditsInBank  Occupation  NoOfDependents  Telephone  \
353              2                  2           3               1          1   
132              2                  1           3               1          2   
639              2                  2           3               1          2   
591              3                  2           4               2          2   
548              2                  1           3               1          1   
487              2                  1           3               1          2   
309              2                  1           2               2          1   
849              3                  1           3               2          2   
798              2                  1           3               1    

In [20]:
GC_X_Test.sample(10)

AccountBalance  CreditDuration  PaymentStatusofPreviousCredit  Purpose  \
844               2              24                              4       10   
989               2              24                              2        0   
255               3              15                              2        9   
698               2              45                              4        1   
958               2              30                              0        9   
231               4               6                              0        3   
710               4              24                              2        0   
545               2               8                              2        9   
672               2              18                              2        3   
643               1              48                              4        1   

     CreditAmount  ValueSavingsOrStocks  CurrentEmploymentDuration  \
844         11938                     1                          3   
989          2718                     1                          3   
255          2687                     1                          4   
698          4576                     2                          1   
958          4280                     2                          3   
231           426                     1                          5   
710          3757                     1                          5   
545           907                     1                          2   
672          1113                     1                          3   
643          6143                     1                          5   

     InstalmentPercentage  SexAndMaritalStatus  Guarantors  \
844                     2                    3           2   
989                     3                    2           1   
255                     2                    3           1   
698                     3                    3           1   
958                     4                    2           1   
231                     4                    4           1   
710                     4                    2           2   
545                     3                    4           1   
672                     4                    2           3   
643                     4                    2           1   

     CurrentAddressDuration  MostValuableAsset  Age(years)  ConcurrentCredits  \
844                       3                  3          39                  3   
989                       4                  2          20                  3   
255                       4                  2          26                  3   
698                       4                  3          27                  3   
958                       4                  3          26                  3   
231                       4                  3          39                  3   
710                       4                  4          62                  3   
545                       2                  1          26                  3   
672                       4                  1          26                  3   
643                       4                  4          58                  2   

     ApartmentType  NoOfCreditsInBank  Occupation  NoOfDependents  Telephone  \
844              2                  2           4               2          2   
989              1                  1           2               1          2   
255              1                  1           3               1          2   
698              2                  1           3               1          1   
958              1                  2           2               1          1   
231              2                  1           2               1          1   
710              3                  1           3               1          2   
545              2                  1           3               1          2   
672              2                  1           2               2    

## 4. Implementation of Predictive Model

The predictive model is implemented using Logistic regression

For this, we need to train our dataset using linearRegression model from sklearn package

In [21]:
#Linera Binary Regression

from sklearn.linear_model import LogisticRegression
clf = LogisticRegression()

In [22]:
clf.fit(GC_X_Train, GC_Y_Train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [23]:
predictions = clf.predict(GC_X_Test)

## 5. Model Evaluvation

In [25]:
from sklearn.metrics import confusion_matrix
clf.score(GC_X_Test, GC_Y_Test)

0.765

In [26]:
confusion_matrix(GC_Y_Test, predictions)

array([[ 28,  40],
       [  7, 125]])

From the above figs, we can see how many true positive, false positive, ture negative and false negative are there. In this, therea are 28 true negative, 40 false negative and 7 false positive and 125 true positive.

In [29]:
y_score = clf.decision_function(GC_X_Test)

from sklearn.metrics import average_precision_score
average_precision = average_precision_score(GC_Y_Test, y_score)

print('Average precision-recall score: {0:0.2f}'.format(
      average_precision))

Average precision-recall score: 0.88
